File: apps/android/src/main/kotlin/com/futureme/financialai/presentation/FutureMeViewModel.kt
package com.futureme.financialai.presentation

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import com.futureme.shared.domain.FutureMeProduct
import com.futureme.shared.models.AssistantPrompt
import com.futureme.shared.models.AnalyticsEvent
import com.futureme.shared.models.BankingVisionDemo
import com.futureme.shared.models.DashboardSnapshot
import com.futureme.shared.models.DecisionJournalEntry
import com.futureme.shared.models.FinancialProfile
import com.futureme.shared.models.FinancialGpsResult
import com.futureme.shared.models.FinancialExplainability
import com.futureme.shared.models.FutureOutcomeContribution
import com.futureme.shared.models.GoalProbabilityResult
import com.futureme.shared.models.Insight
import com.futureme.shared.models.ExecutiveDemoExperience
import com.futureme.shared.models.LifeDecisionSimulation
import com.futureme.shared.models.LifeEventPlan
import com.futureme.shared.models.LifeReadinessResult
import com.futureme.shared.models.LifeTimelinePoint
import com.futureme.shared.models.MoneyLeak
import com.futureme.shared.models.MonthlyFinancialReview
import com.futureme.shared.models.Mission
import com.futureme.shared.models.MissionAnalyticsSnapshot
import com.futureme.shared.models.MissionControlSnapshot
import com.futureme.shared.models.MissionExecutionCenter
import com.futureme.shared.models.NextBestAction
import com.futureme.shared.models.OpportunityRecommendation
import com.futureme.shared.models.ReadinessImprovementPlan
import com.futureme.shared.models.Scenario
import com.futureme.shared.models.ScenarioComparison
import com.futureme.shared.models.ScenarioImpactHeatmap
import com.futureme.shared.models.ScenarioResult
import com.futureme.shared.models.SuggestedQuestion
import com.futureme.shared.models.UserIdentity

enum class FutureMeScreen {
    DASHBOARD,
    BANKING,
    READINESS,
    TIMELINE,
    SCENARIOS,
    SCENARIO_DETAIL,
    COMPARISON,
    LIFE_EVENTS,
    MONEY_LEAKS,
    REVIEW,
    ASSISTANT,
}

data class ChatMessage(
    val id: Long,
    val text: String,
    val isUser: Boolean,
)

data class FutureMeContent(
    val identity: UserIdentity,
    val profile: FinancialProfile,
    val dashboard: DashboardSnapshot,
    val scenarios: List<Scenario>,
    val recentResults: List<ScenarioResult>,
    val insights: List<Insight>,
    val financialGps: FinancialGpsResult,
    val goals: List<GoalProbabilityResult>,
    val lifeEvents: List<LifeEventPlan>,
    val moneyLeaks: List<MoneyLeak>,
    val readiness: List<LifeReadinessResult>,
    val readinessPlans: List<ReadinessImprovementPlan>,
    val decisionSimulations: List<LifeDecisionSimulation>,
    val lifeTimeline: List<LifeTimelinePoint>,
    val executiveDemo: ExecutiveDemoExperience,
    val opportunities: List<OpportunityRecommendation>,
    val nextBestAction: NextBestAction,
    val financialExplainability: FinancialExplainability,
    val scenarioImpactHeatmaps: List<ScenarioImpactHeatmap>,
    val monthlyReviews: List<MonthlyFinancialReview>,
    val decisionJournal: List<DecisionJournalEntry>,
    val futureOutcomeContributions: List<FutureOutcomeContribution>,
    val bankingVisionDemo: BankingVisionDemo,
    val analyticsEvents: List<AnalyticsEvent>,
    val missions: List<Mission>,
    val missionControl: MissionControlSnapshot,
    val missionExecution: MissionExecutionCenter,
    val missionAnalytics: MissionAnalyticsSnapshot,
    val suggestedQuestions: List<SuggestedQuestion>,
    val disclaimer: String,
    val selectedScenario: Scenario? = null,
    val selectedResult: ScenarioResult? = null,
    val comparison: ScenarioComparison,
    val messages: List<ChatMessage> = listOf(
        ChatMessage(
            id = 0L,
            text = "Tell me which mission matters now. I will explain the blocker, timeline, " +
                "and highest-impact action.",
            isUser = false,
        ),
    ),
    val screen: FutureMeScreen = FutureMeScreen.DASHBOARD,
)

sealed interface FutureMeUiState {
    data object Loading : FutureMeUiState
    data class Content(val data: FutureMeContent) : FutureMeUiState
    data class Empty(val message: String) : FutureMeUiState
    data class Error(val message: String) : FutureMeUiState
}

class FutureMeViewModel(
    private val product: FutureMeProduct,
) : ViewModel() {
    var uiState by mutableStateOf<FutureMeUiState>(FutureMeUiState.Loading)
        private set

    init {
        load()
    }

    fun load() {
        uiState = FutureMeUiState.Loading
        uiState = try {
            val bootstrap = product.bootstrap()
            if (bootstrap.scenarios.isEmpty()) {
                FutureMeUiState.Empty("No demo scenarios are available yet.")
            } else {
                FutureMeUiState.Content(
                    FutureMeContent(
                        identity = bootstrap.identity,
                        profile = bootstrap.profile,
                        dashboard = bootstrap.dashboard,
                        scenarios = bootstrap.scenarios,
                        recentResults = bootstrap.recentScenarioResults,
                        insights = bootstrap.insights,
                        financialGps = bootstrap.financialGps,
                        goals = bootstrap.goals,
                        lifeEvents = bootstrap.lifeEvents,
                        moneyLeaks = bootstrap.moneyLeaks,
                        readiness = bootstrap.readiness,
                        readinessPlans = bootstrap.readinessPlans,
                        decisionSimulations = bootstrap.decisionSimulations,
                        lifeTimeline = bootstrap.lifeTimeline,
                        executiveDemo = bootstrap.executiveDemo,
                        opportunities = bootstrap.opportunities,
                        nextBestAction = bootstrap.nextBestAction,
                        financialExplainability = bootstrap.financialExplainability,
                        scenarioImpactHeatmaps = bootstrap.scenarioImpactHeatmaps,
                        monthlyReviews = bootstrap.monthlyReviews,
                        decisionJournal = bootstrap.decisionJournal,
                        futureOutcomeContributions = bootstrap.futureOutcomeContributions,
                        bankingVisionDemo = bootstrap.bankingVisionDemo,
                        analyticsEvents = bootstrap.analyticsEvents,
                        missions = bootstrap.missions,
                        missionControl = bootstrap.missionControl,
                        missionExecution = bootstrap.missionExecution,
                        missionAnalytics = bootstrap.missionAnalytics,
                        suggestedQuestions = bootstrap.suggestedQuestions,
                        disclaimer = bootstrap.disclaimer,
                        comparison = product.compare("move-to-texas", "stay-in-new-jersey"),
                    ),
                )
            }
        } catch (error: Throwable) {
            FutureMeUiState.Error(error.message ?: "FutureMe could not load the demo workspace.")
        }
    }

    fun navigate(screen: FutureMeScreen) {
        when (screen) {
            FutureMeScreen.READINESS -> product.recordAnalyticsEvent("readiness_viewed")
            FutureMeScreen.REVIEW -> product.recordAnalyticsEvent("monthly_review_opened")
            else -> Unit
        }
        updateContent { it.copy(screen = screen) }
    }

    fun openScenario(scenario: Scenario) {
        runCatching {
            product.simulate(scenario.id)
        }.onSuccess { result ->
            product.recordAnalyticsEvent("scenario_created", scenario.id)
            updateContent {
                it.copy(
                    selectedScenario = scenario,
                    selectedResult = result,
                    screen = FutureMeScreen.SCENARIO_DETAIL,
                )
            }
        }.onFailure { error ->
            uiState = FutureMeUiState.Error(
                error.message ?: "The scenario could not be simulated.",
            )
        }
    }

    fun compare(left: Scenario, right: Scenario) {
        runCatching {
            product.compare(left.id, right.id)
        }.onSuccess { comparison ->
            updateContent {
                it.copy(
                    comparison = comparison,
                    screen = FutureMeScreen.COMPARISON,
                )
            }
        }.onFailure { error ->
            uiState = FutureMeUiState.Error(
                error.message ?: "The scenarios could not be compared.",
            )
        }
    }

    fun askAssistant(question: String) {
        val normalized = question.trim()
        if (normalized.isEmpty()) return

        val current = (uiState as? FutureMeUiState.Content)?.data ?: return
        val nextId = (current.messages.maxOfOrNull(ChatMessage::id) ?: 0L) + 1L
        val response = runCatching {
            product.ask(
                AssistantPrompt(
                    question = normalized,
                    latestScenarioId = current.selectedScenario?.id,
                ),
            )
        }.getOrElse { error ->
            updateContent {
                it.copy(
                    messages = it.messages + ChatMessage(
                        id = nextId,
                        text = normalized,
                        isUser = true,
                    ) + ChatMessage(
                        id = nextId + 1L,
                        text = error.message ?: "I could not model that question.",
                        isUser = false,
                    ),
                    screen = FutureMeScreen.ASSISTANT,
                )
            }
            return
        }

        updateContent {
            it.copy(
                messages = it.messages + ChatMessage(
                    id = nextId,
                    text = normalized,
                    isUser = true,
                ) + ChatMessage(
                    id = nextId + 1L,
                    text = response.answer,
                    isUser = false,
                ),
                screen = FutureMeScreen.ASSISTANT,
            )
        }
    }

    fun acceptRecommendation() {
        val current = (uiState as? FutureMeUiState.Content)?.data ?: return
        product.recordAnalyticsEvent(
            "recommendation_accepted",
            current.nextBestAction.recommendationId,
        )
        updateContent { it.copy(analyticsEvents = product.analyticsEvents()) }
    }

    fun acceptMissionAction(actionId: String) {
        val bootstrap = product.completeMissionAction(actionId)
        updateContent {
            it.copy(
                missions = bootstrap.missions,
                missionControl = bootstrap.missionControl,
                missionExecution = bootstrap.missionExecution,
                missionAnalytics = bootstrap.missionAnalytics,
                analyticsEvents = bootstrap.analyticsEvents,
            )
        }
    }

    fun saveDecision() {
        val current = (uiState as? FutureMeUiState.Content)?.data ?: return
        val scenarioId = current.selectedScenario?.id ?: return
        product.saveDecision(scenarioId)
        updateContent {
            it.copy(
                decisionJournal = product.decisionJournal(),
                analyticsEvents = product.analyticsEvents(),
            )
        }
    }

    private fun updateContent(transform: (FutureMeContent) -> FutureMeContent) {
        val current = (uiState as? FutureMeUiState.Content)?.data ?: return
        uiState = FutureMeUiState.Content(transform(current))
    }

    companion object {
        fun factory(product: FutureMeProduct): ViewModelProvider.Factory =
            object : ViewModelProvider.Factory {
                @Suppress("UNCHECKED_CAST")
                override fun <T : ViewModel> create(modelClass: Class<T>): T =
                    FutureMeViewModel(product) as T
            }
    }
}

File: apps/ios/FutureMeFinancial/FutureMeViewModel.swift
import Foundation
import Shared

struct ScenarioCardModel: Identifiable {
    let id: String
    let title: String
    let subtitle: String
    let code: String
    let source: Scenario
}

struct AssistantMessage: Identifiable {
    let id: UUID
    let text: String
    let isUser: Bool
}

struct FutureMeDashboardContent {
    let displayName: String
    let householdName: String
    let dashboard: DashboardSnapshot
    let scenarios: [ScenarioCardModel]
    let selected: ScenarioCardModel
    let result: ScenarioResult
    let comparison: ScenarioComparison
    let insights: [Insight]
    let financialGps: FinancialGpsResult
    let goals: [GoalProbabilityResult]
    let lifeEvents: [LifeEventPlan]
    let moneyLeaks: [MoneyLeak]
    let readiness: [LifeReadinessResult]
    let readinessPlans: [ReadinessImprovementPlan]
    let decisionSimulations: [LifeDecisionSimulation]
    let lifeTimeline: [LifeTimelinePoint]
    let executiveDemo: ExecutiveDemoExperience
    let opportunities: [OpportunityRecommendation]
    let nextBestAction: NextBestAction
    let financialExplainability: FinancialExplainability
    let scenarioImpactHeatmaps: [ScenarioImpactHeatmap]
    let monthlyReviews: [MonthlyFinancialReview]
    let decisionJournal: [DecisionJournalEntry]
    let futureOutcomeContributions: [FutureOutcomeContribution]
    let bankingVisionDemo: BankingVisionDemo
    let analyticsEvents: [AnalyticsEvent]
    let missions: [Mission]
    let missionControl: MissionControlSnapshot
    let missionAnalytics: MissionAnalyticsSnapshot
    let suggestions: [SuggestedQuestion]
    let messages: [AssistantMessage]
    let disclaimer: String
}

enum FutureMeViewState {
    case loading
    case content(FutureMeDashboardContent)
    case empty(String)
    case error(String)
}

@MainActor
final class FutureMeViewModel: ObservableObject {
    @Published private(set) var state: FutureMeViewState = .loading

    private let product: FutureMeProduct
    private let secureStore: SecureStoring

    init(
        product: FutureMeProduct = FutureMeProduct(),
        secureStore: SecureStoring = KeychainSecureStore()
    ) {
        self.product = product
        self.secureStore = secureStore
        load()
    }

    func load() {
        state = .loading

        let bootstrap = product.bootstrap()
        let mappedScenarios = bootstrap.scenarios.map {
            ScenarioCardModel(
                id: $0.id,
                title: $0.title,
                subtitle: $0.subtitle,
                code: Self.code(for: $0),
                source: $0
            )
        }

        guard mappedScenarios.count >= 2 else {
            state = .empty("Add at least two scenarios to unlock comparison mode.")
            return
        }

        let initial = mappedScenarios.first { $0.id == "move-to-texas" } ?? mappedScenarios[0]
        state = .content(
            FutureMeDashboardContent(
                displayName: bootstrap.identity.displayName,
                householdName: bootstrap.identity.householdName,
                dashboard: bootstrap.dashboard,
                scenarios: mappedScenarios,
                selected: initial,
                result: product.simulate(scenarioId: initial.id),
                comparison: product.compare(
                    leftScenarioId: "move-to-texas",
                    rightScenarioId: "stay-in-new-jersey"
                ),
                insights: bootstrap.insights,
                financialGps: bootstrap.financialGps,
                goals: bootstrap.goals,
                lifeEvents: bootstrap.lifeEvents,
                moneyLeaks: bootstrap.moneyLeaks,
                readiness: bootstrap.readiness,
                readinessPlans: bootstrap.readinessPlans,
                decisionSimulations: bootstrap.decisionSimulations,
                lifeTimeline: bootstrap.lifeTimeline,
                executiveDemo: bootstrap.executiveDemo,
                opportunities: bootstrap.opportunities,
                nextBestAction: bootstrap.nextBestAction,
                financialExplainability: bootstrap.financialExplainability,
                scenarioImpactHeatmaps: bootstrap.scenarioImpactHeatmaps,
                monthlyReviews: bootstrap.monthlyReviews,
                decisionJournal: bootstrap.decisionJournal,
                futureOutcomeContributions: bootstrap.futureOutcomeContributions,
                bankingVisionDemo: bootstrap.bankingVisionDemo,
                analyticsEvents: bootstrap.analyticsEvents,
                missions: bootstrap.missions,
                missionControl: bootstrap.missionControl,
                missionAnalytics: bootstrap.missionAnalytics,
                suggestions: bootstrap.suggestedQuestions,
                messages: [
                    AssistantMessage(
                        id: UUID(),
                        text: "Tell me which mission matters now. I will explain the blocker, timeline, and highest-impact action.",
                        isUser: false
                    ),
                ],
                disclaimer: bootstrap.disclaimer
            )
        )

        // Stores only an opaque demo profile identifier. Financial values remain in memory.
        try? secureStore.put(bootstrap.profile.profileId, for: "active-demo-profile")
    }

    func select(_ scenario: ScenarioCardModel) {
        guard case let .content(content) = state else {
            return
        }

        _ = product.recordAnalyticsEvent(typeCode: "scenario_created", subjectId: scenario.id)
        state = .content(
            FutureMeDashboardContent(
                displayName: content.displayName,
                householdName: content.householdName,
                dashboard: content.dashboard,
                scenarios: content.scenarios,
                selected: scenario,
                result: product.simulate(scenarioId: scenario.id),
                comparison: content.comparison,
                insights: content.insights,
                financialGps: content.financialGps,
                goals: content.goals,
                lifeEvents: content.lifeEvents,
                moneyLeaks: content.moneyLeaks,
                readiness: content.readiness,
                readinessPlans: content.readinessPlans,
                decisionSimulations: content.decisionSimulations,
                lifeTimeline: content.lifeTimeline,
                executiveDemo: content.executiveDemo,
                opportunities: content.opportunities,
                nextBestAction: content.nextBestAction,
                financialExplainability: content.financialExplainability,
                scenarioImpactHeatmaps: content.scenarioImpactHeatmaps,
                monthlyReviews: content.monthlyReviews,
                decisionJournal: content.decisionJournal,
                futureOutcomeContributions: content.futureOutcomeContributions,
                bankingVisionDemo: content.bankingVisionDemo,
                analyticsEvents: product.analyticsEvents(),
                missions: content.missions,
                missionControl: content.missionControl,
                missionAnalytics: content.missionAnalytics,
                suggestions: content.suggestions,
                messages: content.messages,
                disclaimer: content.disclaimer
            )
        )
    }

    func compare(_ left: ScenarioCardModel, _ right: ScenarioCardModel) {
        guard left.id != right.id, case let .content(content) = state else {
            return
        }

        state = .content(
            FutureMeDashboardContent(
                displayName: content.displayName,
                householdName: content.householdName,
                dashboard: content.dashboard,
                scenarios: content.scenarios,
                selected: content.selected,
                result: content.result,
                comparison: product.compare(
                    leftScenarioId: left.id,
                    rightScenarioId: right.id
                ),
                insights: content.insights,
                financialGps: content.financialGps,
                goals: content.goals,
                lifeEvents: content.lifeEvents,
                moneyLeaks: content.moneyLeaks,
                readiness: content.readiness,
                readinessPlans: content.readinessPlans,
                decisionSimulations: content.decisionSimulations,
                lifeTimeline: content.lifeTimeline,
                executiveDemo: content.executiveDemo,
                opportunities: content.opportunities,
                nextBestAction: content.nextBestAction,
                financialExplainability: content.financialExplainability,
                scenarioImpactHeatmaps: content.scenarioImpactHeatmaps,
                monthlyReviews: content.monthlyReviews,
                decisionJournal: content.decisionJournal,
                futureOutcomeContributions: content.futureOutcomeContributions,
                bankingVisionDemo: content.bankingVisionDemo,
                analyticsEvents: content.analyticsEvents,
                missions: content.missions,
                missionControl: content.missionControl,
                missionAnalytics: content.missionAnalytics,
                suggestions: content.suggestions,
                messages: content.messages,
                disclaimer: content.disclaimer
            )
        )
    }

    func ask(_ question: String) {
        let normalized = question.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !normalized.isEmpty, case let .content(content) = state else {
            return
        }

        let response = product.ask(
            prompt: AssistantPrompt(
                question: normalized,
                latestScenarioId: content.selected.id
            )
        )
        let messages = content.messages + [
            AssistantMessage(id: UUID(), text: normalized, isUser: true),
            AssistantMessage(id: UUID(), text: response.answer, isUser: false),
        ]

        state = .content(
            FutureMeDashboardContent(
                displayName: content.displayName,
                householdName: content.householdName,
                dashboard: content.dashboard,
                scenarios: content.scenarios,
                selected: content.selected,
                result: content.result,
                comparison: content.comparison,
                insights: content.insights,
                financialGps: content.financialGps,
                goals: content.goals,
                lifeEvents: content.lifeEvents,
                moneyLeaks: content.moneyLeaks,
                readiness: content.readiness,
                readinessPlans: content.readinessPlans,
                decisionSimulations: content.decisionSimulations,
                lifeTimeline: content.lifeTimeline,
                executiveDemo: content.executiveDemo,
                opportunities: content.opportunities,
                nextBestAction: content.nextBestAction,
                financialExplainability: content.financialExplainability,
                scenarioImpactHeatmaps: content.scenarioImpactHeatmaps,
                monthlyReviews: content.monthlyReviews,
                decisionJournal: content.decisionJournal,
                futureOutcomeContributions: content.futureOutcomeContributions,
                bankingVisionDemo: content.bankingVisionDemo,
                analyticsEvents: product.analyticsEvents(),
                missions: content.missions,
                missionControl: content.missionControl,
                missionAnalytics: content.missionAnalytics,
                suggestions: content.suggestions,
                messages: messages,
                disclaimer: content.disclaimer
            )
        )
    }

    func acceptRecommendation() {
        guard case let .content(content) = state else {
            return
        }
        _ = product.recordAnalyticsEvent(
            typeCode: "recommendation_accepted",
            subjectId: content.nextBestAction.recommendationId
        )
        replace(content, analyticsEvents: product.analyticsEvents())
    }

    func acceptMissionAction(_ actionId: String) {
        guard case let .content(content) = state else {
            return
        }
        _ = product.recordAnalyticsEvent(
            typeCode: "mission_action_completed",
            subjectId: actionId
        )
        replace(content, analyticsEvents: product.analyticsEvents())
    }

    func saveDecision() {
        guard case let .content(content) = state else {
            return
        }
        _ = product.saveDecision(scenarioId: content.selected.id)
        replace(
            content,
            decisionJournal: product.decisionJournal(),
            analyticsEvents: product.analyticsEvents()
        )
    }

    private func replace(
        _ content: FutureMeDashboardContent,
        decisionJournal: [DecisionJournalEntry]? = nil,
        analyticsEvents: [AnalyticsEvent]? = nil
    ) {
        state = .content(
            FutureMeDashboardContent(
                displayName: content.displayName,
                householdName: content.householdName,
                dashboard: content.dashboard,
                scenarios: content.scenarios,
                selected: content.selected,
                result: content.result,
                comparison: content.comparison,
                insights: content.insights,
                financialGps: content.financialGps,
                goals: content.goals,
                lifeEvents: content.lifeEvents,
                moneyLeaks: content.moneyLeaks,
                readiness: content.readiness,
                readinessPlans: content.readinessPlans,
                decisionSimulations: content.decisionSimulations,
                lifeTimeline: content.lifeTimeline,
                executiveDemo: content.executiveDemo,
                opportunities: content.opportunities,
                nextBestAction: content.nextBestAction,
                financialExplainability: content.financialExplainability,
                scenarioImpactHeatmaps: content.scenarioImpactHeatmaps,
                monthlyReviews: content.monthlyReviews,
                decisionJournal: decisionJournal ?? content.decisionJournal,
                futureOutcomeContributions: content.futureOutcomeContributions,
                bankingVisionDemo: content.bankingVisionDemo,
                analyticsEvents: analyticsEvents ?? content.analyticsEvents,
                missions: content.missions,
                missionControl: content.missionControl,
                missionAnalytics: content.missionAnalytics,
                suggestions: content.suggestions,
                messages: content.messages,
                disclaimer: content.disclaimer
            )
        )
    }

    private static func code(for scenario: Scenario) -> String {
        switch scenario.id {
        case "buy-home", "wait-to-buy": "HOME"
        case "refinance-now", "keep-current-loan": "REFI"
        case "pay-off-cards": "DEBT"
        case "move-to-texas", "stay-in-new-jersey": "MOVE"
        case "have-a-child": "FAM"
        case "spouse-stops-working": "ONE"
        case "start-business": "BIZ"
        case "invest-more": "INV"
        default: "SAFE"
        }
    }
}
